awscli/customizations/dynamodb.py

```diff
@@ -0,0 +1,49 @@
+# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import base64
+import binascii
+import logging
+
+from awscli.compat import six
+
+logger = logging.getLogger(__name__)
+
+
+def register_dynamodb_paginator_fix(event_emitter):
+    event_emitter.register(
+        'calling-command.dynamodb.*', _maybe_register_pagination_fix
+    )
+
+
+def _maybe_register_pagination_fix(parsed_globals, session, **kwargs):
+    if parsed_globals.paginate:
+        session.register(
+            'after-call.dynamodb.*', parse_last_evaluated_key_binary
+        )
+
+
+def parse_last_evaluated_key_binary(parsed, **kwargs):
+    # Because we disable parsing blobs into a binary type and leave them as
+    # a base64 string if a binary field is present in the continuation token
+    # as is the case with dynamodb the binary will be double encoded. This
+    # ensures that the continuation token is properly converted to binary to
+    # avoid double encoding the contination token.
+    last_evaluated_key = parsed.get('LastEvaluatedKey', None)
+    if last_evaluated_key is None:
+        return
+    for key, val in last_evaluated_key.items():
+        if 'B' in val:
+            try:
+                val['B'] = base64.b64decode(val['B'])
+            except (binascii.Error, TypeError):
```

Comment:
Thoughts on a hard failure here?  If this decode fails that indicates we received bad data from the service, or something else mucked with this value.  Either way, we can't guarantee that we're paginating correctly which brings us back to the inconsistent/complete pagination results.

FWIW if the base64 decode fails in the botocore parser layer (i.e we didn't mess with the blob parser defaults), this would propagate an exception.

Reply:
Hmmm, yeah. I think hard failing is probably better than silently producing wrong results (potentially). This was mostly defensive in case something else ever parsed this into bytes already and the fact that it's ambiguous on Python 2. Though in that case one of the tests will probably fail so I think we can hard fail.